File: listenmoe-api/src/main/java/me/echeung/listenmoeapi/models/Song.java
package me.echeung.listenmoeapi.models;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.List;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import me.echeung.listenmoeapi.APIClient;

@Getter
@EqualsAndHashCode
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Song implements Parcelable {
    private int id;
    private String title;
    private String titleRomaji;
    private String titleSearchRomaji;
    private List<SongDescriptor> albums;
    private List<SongDescriptor> artists;
    private List<SongDescriptor> sources;
    private List<String> groups;
    private List<String> tags;
    private String notes;
    private int duration;
    private boolean enabled;
    private User uploader;

    @Setter
    private boolean favorite;

    public String getAlbumsString() {
        StringBuilder s = new StringBuilder();
        if (albums != null) {
            for (SongDescriptor album : albums) {
                if (s.length() != 0)
                    s.append(", ");
                s.append(album.getName());
            }
        }
        return s.toString();
    }

    public String getArtistsString() {
        StringBuilder s = new StringBuilder();
        if (artists != null) {
            for (SongDescriptor artist : artists) {
                if (s.length() != 0)
                    s.append(", ");
                s.append(artist.getName());
            }
        }
        return s.toString();
    }

    public String getSourcesString() {
        StringBuilder s = new StringBuilder();
        if (artists != null) {
            for (SongDescriptor source : sources) {
                if (s.length() != 0)
                    s.append(", ");
                s.append(source.getName());
            }
        }
        return s.toString();
    }

    @Override
    public String toString() {
        return String.format("%s - %s", getTitle(), getArtistsString());
    }

    public String getAlbumArtUrl() {
        if (!albums.isEmpty()) {
            for (SongDescriptor album : albums) {
                if (album.getImage() != null) {
                    return APIClient.CDN_ALBUM_ART_URL + album.getImage();
                }
            }
        }

        return null;
    }

    public boolean search(String query) {
        query = query.toLowerCase().trim();

        if (title != null && title.toLowerCase().contains(query)) {
            return true;
        }

        if (titleRomaji != null && titleRomaji.toLowerCase().contains(query)) {
            return true;
        }

        if (titleSearchRomaji != null && titleSearchRomaji.toLowerCase().contains(query)) {
            return true;
        }

        if (albums != null) {
            for (SongDescriptor album : albums) {
                if (album.getName() != null && album.getName().toLowerCase().contains(query)) {
                    return true;
                }
                if (album.getNameRomaji() != null && album.getNameRomaji().toLowerCase().contains(query)) {
                    return true;
                }
            }
        }

        if (artists != null) {
            for (SongDescriptor artist : artists) {
                if (artist.getName() != null && artist.getName().toLowerCase().contains(query)) {
                    return true;
                }
                if (artist.getNameRomaji() != null && artist.getNameRomaji().toLowerCase().contains(query)) {
                    return true;
                }
            }
        }

        if (groups != null) {
            for (String group : groups) {
                if (group != null && group.toLowerCase().contains(query)) {
                    return true;
                }
            }
        }

        if (tags != null) {
            for (String tag : tags) {
                if (tag != null && tag.toLowerCase().contains(query)) {
                    return true;
                }
            }
        }

        return false;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeInt(id);
        parcel.writeString(title);
        parcel.writeString(titleRomaji);
        parcel.writeString(titleSearchRomaji);
        parcel.writeTypedList(albums);
        parcel.writeTypedList(artists);
        parcel.writeTypedList(sources);
        parcel.writeStringList(groups);
        parcel.writeStringList(tags);
        parcel.writeString(notes);
        parcel.writeInt(duration);
        parcel.writeByte(enabled ? (byte) 1 : 0);
        parcel.writeParcelable(uploader, 0);
        parcel.writeByte(favorite ? (byte) 1 : 0);
    }

    public Song(Parcel in) {
        this.id = in.readInt();
        this.title = in.readString();
        this.titleRomaji = in.readString();
        this.titleSearchRomaji = in.readString();
        in.readTypedList(this.albums, SongDescriptor.CREATOR);
        in.readTypedList(this.artists, SongDescriptor.CREATOR);
        in.readTypedList(this.sources, SongDescriptor.CREATOR);
        in.readStringList(this.groups);
        in.readStringList(this.tags);
        this.notes = in.readString();
        this.duration = in.readInt();
        this.enabled = in.readByte() == 1;
        this.uploader = in.readParcelable(User.class.getClassLoader());
        this.favorite = in.readByte() == 1;
    }

    public static final Creator CREATOR = new Creator() {
        public Song createFromParcel(Parcel in) {
            return new Song(in);
        }

        public Song[] newArray(int size) {
            return new Song[size];
        }
    };

}

File: app/src/main/kotlin/me/echeung/moemoekyun/client/api/model/Song.kt
package me.echeung.moemoekyun.client.api.model

import kotlinx.serialization.Serializable
import me.echeung.moemoekyun.client.api.Library
import me.echeung.moemoekyun.util.SongFormatter
import org.koin.core.component.KoinComponent
import org.koin.core.component.get
import java.util.Locale

@Serializable
data class Song(
    var id: Int = 0,
    var title: String? = null,
    var titleRomaji: String? = null,
    var artists: List<SongDescriptor>? = null,
    var sources: List<SongDescriptor>? = null,
    var albums: List<SongDescriptor>? = null,
    var duration: Int = 0,
    var enabled: Boolean = false,
    var favorite: Boolean = false,
) : KoinComponent {

    fun getTitleString(): String {
        val songFormatter: SongFormatter = get()
        return songFormatter.getTitle(this)
    }

    fun getArtistsString(): String {
        val songFormatter: SongFormatter = get()
        return songFormatter.getArtists(this)
    }

    fun getAlbumsString(): String {
        val songFormatter: SongFormatter = get()
        return songFormatter.getAlbums(this)
    }

    fun getSourcesString(): String {
        val songFormatter: SongFormatter = get()
        return songFormatter.getSources(this)
    }

    val durationString: String
        get() {
            var minutes = (duration / 60).toLong()
            val seconds = (duration % 60).toLong()
            return if (minutes < 60) {
                String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds)
            } else {
                val hours = minutes / 60
                minutes %= 60
                String.format(Locale.getDefault(), "%02d:%02d:%02d", hours, minutes, seconds)
            }
        }

    val albumArtUrl: String?
        get() {
            val album = albums?.firstOrNull { it.image != null }
            if (album != null) {
                return Library.CDN_ALBUM_ART_URL + album.image
            }

            return null
        }

    override fun toString(): String {
        val songFormatter: SongFormatter = get()
        return "${songFormatter.getTitle(this)} - ${songFormatter.getArtists(this)}"
    }

    fun search(query: String?): Boolean {
        if (query.isNullOrBlank()) {
            return true
        }

        return title?.contains(query, ignoreCase = true) ?: false ||
            titleRomaji?.contains(query, ignoreCase = true) ?: false ||
            artists?.any { it.contains(query) } ?: false ||
            albums?.any { it.contains(query) } ?: false ||
            sources?.any { it.contains(query) } ?: false
    }
}

fun List<Song>.search(query: String?): List<Song> {
    return asSequence()
        .filter { it.search(query) }
        .toList()
}
